tests: Create tests with bad decorators

AbdealiLoKo opened the pull request:
Bad decorators are decorators that are incomplete. They
don't use functools.wraps and so some important metadata that
we use if not copied into the decorated function.

Related to https://github.com/AbdealiJK/pytest-attrib/issues/1

AbdealiLoKo committed:
tests: Create tests with bad decorators

Bad decorators are decorators that are incomplete. They
don't use functools.wraps and so some important metadata that
we use if not copied into the decorated function.

Related to https://github.com/AbdealiJK/pytest-attrib/issues/1

AbdealiJK commented:
We just add an xfail test so that we know it is a problem.
Not sure how to solve it as of yet :/

Decorator writers need to write these things correctly

jayvdb commented:
I suspect that it is impossible to have a unit test that is a function with `<locals>` in the name, in which case you can detect it and generate an error, telling the coder to fix it.

AbdealiLoKo committed:
pytest_attrib: Use the `parent` provided by pytest

Rather than meddling with the metadata ourselves, we now use the
`parent` attribute provided by pytest. This is much more
reliable as pytest searches for the testcases in a top
to bottom way (I.e. first find modules, then classes, then functions)
and so it becomes easy for it to track the heirarchy.

AbdealiJK commented:
After digging around, I found that pytest already has the heirarchy of the tests. (As it collects tests for each class separately).

This is much more reliable as pytest does: module -> class -> function, and has all the metadata stored for it. i've added a commit that uses this now, so it's more reliable + fixes the xfailing testcase for https://github.com/AbdealiJK/pytest-attrib/issues/3 too !
